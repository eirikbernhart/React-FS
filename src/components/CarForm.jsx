import React, { Component } from 'react';
import './CarForm.css'

class CarForm extends Component {
    constructor(props) {
        super(props);
        this.state = {
            _id: '',
            name: '',
            price: 1234,
            __v: 0
        };
        this.handleUserInput = this.handleUserInput.bind(this);
        this.writeNewCar = this.writeNewCar.bind(this);
    }

    //When the user input changes, set the newCarContent to the value of what's in the inputbox
    handleUserInput(e) {
        this.setState({
            name: e.target.value, //The value of the text input.
        })
    }

    writeNewCar() {
        // call a method that sets the noteContent for a note to the value of the input.
        //ID needs to be 24 chars long
        var randomID = this.scientificToDecimal((Math.floor(1000 + Math.random() * 900000000000000000000000)))
        console.log(randomID)

        this.state._id = randomID.toString();
        this.props.addCar(this.state);
       
        
    }

    //Implementation: https://gist.github.com/jiggzson/b5f489af9ad931e3d186 
    scientificToDecimal(num) {
    //if the number is in scientific notation remove it
    if(/\d+\.?\d*e[\+\-]*\d+/i.test(num)) {
        var zero = '0',
            parts = String(num).toLowerCase().split('e'), //split into coeff and exponent
            e = parts.pop(),//store the exponential part
            l = Math.abs(e), //get the number of zeros
            sign = e/l,
            coeff_array = parts[0].split('.');
        if(sign === -1) {
            num = zero + '.' + new Array(l).join(zero) + coeff_array.join('');
        }
        else {
            var dec = coeff_array[1];
            if(dec) l = l - dec.length;
            num = coeff_array.join('') + new Array(l+1).join(zero);
        }
    }
    return num;
};

    render() {
        return(
            <div className="formWrapper">
                <input className="noteInput"
                placeholder="Add a car to wishlist" 
                value={this.state.name}
                onChange={this.handleUserInput} />
                <button className="noteButton"
                onClick={this.writeNewCar}>Add Car</button>
            </div>
        )
    }


}

export default CarForm;