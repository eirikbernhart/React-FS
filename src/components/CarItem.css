
.itemContent {
    padding:10px 10px 38px 34px;
    height: auto; 
    margin: auto;
    margin-bottom: 8px;
    width: 60%;
    font-family: 'Lato', sans-serif;
    font-size: 1.4em;
    color: #222;
    background-color: #eed; 
    border-radius: 3px;
    box-shadow: 43x 3px 2px #333;
    opacity: 1;
  }
  
  .closebtn {
      color: #AA9;
      font-weight: bold;
      float: right;
      cursor: pointer;
      font-size: 1.3em;
  }