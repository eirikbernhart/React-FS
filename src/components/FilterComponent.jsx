import React, { Component } from 'react';
import './FilterComponent.css'
import { setTextFilter } from '../actions/filters';
import { connect } from 'react-redux';



class FilterComponent extends Component {
    constructor(props) {
        super(props);
        this.state = {
            search: ''
        };
    }

    render() {
        return(
            <div>
                SÃ¸k
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={(e) => {
                        this.props.dispatch(setTextFilter(e.target.value));
                    }}></input>
            </div>
        )
    }


}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

export default connect(mapStateToProps)(FilterComponent);