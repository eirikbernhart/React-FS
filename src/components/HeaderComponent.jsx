import React, { Component } from 'react';
import catPic from '.././resources/img/404NotFound.jpeg';
import { Link, NavLink } from 'react-router-dom';
import '.././styles/styles.scss';

  



class HeaderComponent extends Component {

    
     render() {
         return (
            <div>
                <header style={{backgroundColor: "white"}}>
                    <p>I'm da HEADER!</p>
                    <NavLink to="/" activeClassName="is-active" exact={true}>Login-page</NavLink>
                    <NavLink to="/register" activeClassName="is-active" exact={true}>Register-page</NavLink>
                    <NavLink to="/main" activeClassName="is-active" exact={true}>Main-page</NavLink>
                </header>              
            </div>
         )
    }


}


export default HeaderComponent;