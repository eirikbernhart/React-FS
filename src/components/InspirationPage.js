import React, { Component } from 'react';
import { Table, Button } from 'reactstrap';

class InspirationPage extends Component {
    
    render() {
        return(
            <div className="App-header">
                <h2>Inspiration page</h2>
            </div>   
           
        )
    }
}

export default InspirationPage;