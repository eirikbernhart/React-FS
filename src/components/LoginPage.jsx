import React, { Component } from 'react';
import { Button, Form, FormGroup, Label, Input } from 'reactstrap';
import 'bootstrap/dist/css/bootstrap.css';
import { authenticateUser } from '../actions/auth';
import { connect } from 'react-redux';



const urlUser = "http://localhost:1234/users";
const jwt = require('jwt-simple');



export class LoginPage extends Component {
    constructor(props) {
        super(props);
        this.state = {
            username: '',
            password: ''
        }

        this.loginRedux = this.loginRedux.bind(this); //REDUX
        this.handleUsername = this.handleUsername.bind(this);
        this.handlePassword = this.handlePassword.bind(this);
    }

    
    loginRedux() {

        const secret = 'topsecret';

        const urlUserAuth = "http://localhost:1234/usersAuth";

        const body = {
            username: this.state.username,
            password: this.state.password
        }

        fetch(urlUserAuth, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'

            },
            body: JSON.stringify(body)
        })
        .then(res => {
            if(res.status === 401) {
                console.log("NO SUCH USER FROM CLIENT!")
                window.alert(res.statusText);
                throw 0; 
            } else {
                console.log("USER FOUND, YEY!")
                return res.text()
            }
        })
        .then(token => {
            localStorage.token = token
            const user = jwt.decode(localStorage.token, secret)

            console.log("Token from client->login " + token)

            if(localStorage.token) {
                this.props.dispatch(authenticateUser({
                    username: user.username, 
                    authenticated: true
                }));
                this.props.history.push("/main");
            }
        });
    }

    handleUsername(e) {
        this.setState({
            username: e.target.value
        })
    }

    handlePassword(e) {
        this.setState({
            password: e.target.value
        })
    }



     render() {
         return (
            <div>
                <div className="App-header">
                    <h2>Login</h2>
                </div>    
                <Form>
                    <FormGroup>
                        <Label for="userName"><strong>Username</strong></Label>
                        <Input 
                            type="username" 
                            name="username" 
                            id="userName" 
                            placeholder="Username..."
                            value={this.state.username}
                            onChange={this.handleUsername}  
                        />
                    </FormGroup>
                    <FormGroup>
                        <Label for="examplePassword"><strong>Password</strong></Label>
                        <Input 
                            type="password" 
                            name="password" 
                            id="examplePassword" 
                            placeholder="Password..."
                            value={this.state.password}  
                            onChange={this.handlePassword} 
                        />
                    </FormGroup>
                    <Button 
                        color="danger" 
                        onClick={this.loginRedux}>Log in
                    </Button>  
                </Form>
            </div>
         )
    }


    
}



export default connect()(LoginPage);