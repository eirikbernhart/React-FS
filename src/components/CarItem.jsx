import React, { Component } from 'react';
import './CarItem.css';
import { connect } from 'react-redux';
import { makePublic } from '.././actions/cars';
import { makePrivate } from '.././actions/cars';




class CarItem extends Component {

    constructor(props) {
        super(props);
        this.id = props.car._id;
        this.carName = props.car.name;
        this.handleRemoveCar = this.handleRemoveCar.bind(this);
        this.handleCheckBox = this.handleCheckBox.bind(this);
        this.findIndex = this.findIndex.bind(this);

       

        

    }

    componentWillMount() {
        //console.log("Comp mount ran?");
        //console.log("IS IT CHECKED AT START: " + JSON.stringify(this.props.car.isPublic));
        //console.log("WHAT ARE PROPS.CARS: " + JSON.stringify(this.props.cars));
        //props.car retrieves car different car obj twice, props.cars retrives an arr of car obj. FUNKY AF! HUH!
        
            this.setState({
                checked: this.props.car.isPublic
            })
        
    }

    handleRemoveCar(id) {
        console.log("DELETE: HVA ER ID: " + id);
        this.props.removeCar(id);
    }

    handleCheckBox(e) {
        const target = e.target;
        let carID = this.props.car.id;

        if(!this.props.car.isPublic) {
            let index = this.findIndex(carID);
            this.props.dispatch(makePublic({index: index}));
            this.props.makePublicOrPrivate(carID);
        }
        if(this.props.car.isPublic) {
            let index = this.findIndex(carID);
            this.props.dispatch(makePrivate({index: index}));
            this.props.makePublicOrPrivate(carID);
        }
    }

    findIndex(carID){
        var indexVal;
        this.props.cars.map((currentItem, index) => {
            if(currentItem.id === carID) {            
                indexVal = index;                       
            } 
        }); 
        return indexVal;
    }

    render(props) {
        return (
            <div className="itemContent">
                {JSON.stringify(this.props.car.carName)}
                <div className="checkBox">
                  
                <input 
                     name="isGoing"
                     type="checkbox"
                     checked={this.props.car.isPublic}
                     onChange={this.handleCheckBox}/>
                     Public
                </div>

                <span className="closebtn"
                    onClick={() => this.handleRemoveCar(this.props.car.id)}>
                    &times;
            </span>
            </div>
        );
    }
}

function mapStateToProps(state) {
        return {
            cars: state.cars
        };
};

export default connect(mapStateToProps)(CarItem);