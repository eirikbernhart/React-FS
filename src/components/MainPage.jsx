import React, { Component } from 'react';
import { Table } from 'reactstrap';
import CarItem from './CarItem';
import CarForm from './CarForm';
import FilterComponent from './FilterComponent';
import { Button } from 'reactstrap';
import configureStore from '.././store/configureStore';
import { addCar } from '.././actions/cars';
import { removeCar } from '.././actions/cars';
import { clearCar } from '.././actions/cars';
import { searchCar } from '.././actions/cars';
import { connect } from 'react-redux';
import { authenticateUser } from '../actions/auth';
import { unAuthenticateUser } from '../actions/auth';
import { pageVisited } from '../actions/auth';
import selectedCars from '../selectors/cars';
import { NavLink } from 'react-router-dom';
import catPic2 from '.././resources/img/401Unauthorized.jpeg';
import { subscribeToTimer } from '../socket';
import { pushToServer } from '../socket';
import io from 'socket.io-client';
import { setSocket } from '.././actions/socket';










//const url = 'ws://localhost:1234';//Web-Socket related
const socketUrl = 'http://localhost:1234';
const socket = null;


const urlAutomobiles = "http://localhost:1234/automobiles" //Needs token
const jwt = require('jwt-simple');
const secret = 'topsecret';



class MainPage extends Component {

    constructor(props) {
        super(props);

        this.addAutomobileRedux = this.addAutomobileRedux.bind(this); //REDUX
        this.removeCarRedux = this.removeCarRedux.bind(this); //REDUX
        this.getAutomobilesRedux = this.getAutomobilesRedux.bind(this); //REDUX
        this.logoutRedux = this.logoutRedux.bind(this); //REDUX
        //this.socketEvent = this.socketEvent.bind(this);
        //this.initSocket = this.initSocket.bind(this);

        this.state = {
            cars: [],
            socket: null,
            //webSocket: [], //SOCKET: WebSocket-related
            timestamp: 'no timestamp yet',
            testObject: 'empty', //SOCKET: Socket.IO-related
            socketTestData: '', //SOCKET: Socket.IO-related
        };

        
    }

    //initSocket = () => { //ORIGINAL
    /*initSocket() {
        this.socket = io(socketUrl);
        this.socket.on('connect', () => {
            console.log(`User: ${this.props.auth.username} has connected, from MainPage`);
            this.props.dispatch(setSocket({inSocket: this.socket}));

            this.props.activeSocket.inSocket.on('testEvent', () => this.setState({socketTestData: 'TestData'})); //ORIGINAL            
        });
    }*/

    componentWillMount() {

        if(this.props.auth.authenticated && !this.props.auth.visited) {

            //WEBSOCKET: Socket.IO related TRY 
            //this.initSocket();

            this.getAutomobilesRedux(); //REDUX
            this.props.dispatch(pageVisited({visited: true}));

            
        }
    }

    componentDidMount() {
        
    }

    componentWillUnmount() {
        //localStorage.token = '';
    }

    getAutomobilesRedux() {
        const user = jwt.decode(localStorage.token, secret)
        const url = "http://localhost:1234/automobiles/cars" 

        fetch(urlAutomobiles, {
          headers: {
            'X-Token': localStorage.token,
          }
        })
        .then(res => {
          if(res.status === 200) { //NEEDS THIS CHECK BECAUSE FETCH FAILS WHILE USER IS UNAUTHENTICATED!
            return res = res.json();            
          }
        })
        .then(res => {
            var userSpesificData = res.filter(x => x.owner === user.username)

            for(var i = 0; i < userSpesificData.length; i++) {
                this.props.dispatch(addCar({
                    carName: userSpesificData[i].name,
                    id: userSpesificData[i]._id,
                    isPublic: userSpesificData[i].isPublic
                }))
            }
        })
    }

    addAutomobileRedux(automobile) {
            fetch(urlAutomobiles, {
                method: "POST",
                headers: new Headers({
                'Content-Type': 'application/json',
                Accept: 'application/json',
        }),
        body: JSON.stringify(automobile)
        })
        .then(res => {
            this.props.dispatch(addCar({
                carName: automobile.name, 
                id: automobile._id, 
                isPublic: automobile.isPublic
            }));
            console.log("Ble staten oppdatert: " + JSON.stringify(this.state.cars));
        })
        .catch(err => document.write(err));
    }

    removeCarRedux(id) {

        fetch(`${urlAutomobiles}/${id}`, {
        method: "DELETE",
        headers: new Headers({
            'Content-Type': 'application/json',
            Accept: 'application/json',
        }),
        })
        .then(res => {
            this.props.dispatch(removeCar({id: id}))
        })
        .catch(err => document.write(err));
    }

    makePublicOrPrivate(id){

        //this.socketEvent();// DOESNT WORK FOR SOME REASON...

        fetch(`${urlAutomobiles}/${id}`, {
            method: "PUT",
            headers: new Headers({
                'Content-Type': 'application/json',
                Accept: 'application/json',
            }),
            })
            .then(res => {
                
            })
            .catch(err => document.write(err));
            
    }


    logoutRedux() {
        this.props.dispatch(unAuthenticateUser({userName: '', authenticated: false}));
        this.props.dispatch(clearCar());
        localStorage.token = '';
        this.props.history.push("/");
    }

    /*socketEvent() {
        this.props.activeSocket.inSocket.emit('testEvent');
    }*/

    


     render() {
       
        let component;

        if(this.props.auth.authenticated) { 
            component = (
                
                    <div className="App">
                        <div className="App-header">
                            <h2>Car wishlist</h2>
                        </div>
                        <h1>Logged in as: {this.props.auth.username}</h1>

                        <Table bordered>
                            <thead>
                                <tr>
                                    <th>Legg til</th>
                                    <th>Filtrer</th>
                                </tr>
                            </thead>

                            <tbody>
                                <tr>
                                    <th scope="row"><CarForm addCar={this.addAutomobileRedux} /></th>
                                    <th scope="row"><FilterComponent
                                    />
                                    </th>
                                </tr>

                            </tbody>
                        </Table>

                        <Table bordered>
                            <thead>
                                <tr>
                                    <th>#</th>
                                    <th>Car</th>
                                </tr>
                            </thead>

                            <tbody>
                                {this.props.cars.map((currentCar, index) => (
                                    <tr key = {currentCar.id}>
                                        <th scope="row">{index + 1}</th>
                                        <td>
                                        <CarItem
                                            removeCar={this.removeCarRedux}
                                            makePublicOrPrivate={this.makePublicOrPrivate}
                                            car={currentCar} />
                                        </td>
                                    </tr>
                                ))}
                            </tbody>
                        </Table>
                        <Button 
                            color="danger" 
                            onClick={this.logoutRedux}>Log out
                        </Button>  
                        <Button 
                            color="danger" 
                            onClick={this.socketEvent}>SocketEvent
                        </Button>  
                        
                </div>
            )
        } else {
            component = (
                <div className="App-header">
                    <h2>Not authorized, please 
                        <NavLink to="/" activeClassName="is-active" className="linkStyle" exact={true}>login</NavLink>
                    !</h2>
                    <img src= { catPic2 }/>                
                    
                </div> 
            )
        }


         return (
            <div>
                {component}
            </div>
         )
    }


}

function mapStateToProps(state) {
        return {
            cars: selectedCars(state.cars, state.filters),
            auth: state.auth,
            activeSocket: state.activeSocket
        };
};


export default connect(mapStateToProps)(MainPage);








//WEBSOCKET: Socket.IO related TRY 1
        /*subscribeToTimer((err, timestamp) => this.setState({ 
            timestamp: timestamp
        }));*/

        /*pushToServer((err, data) => this.setState({
            testObject: data,
        }));
        console.log("DATA FROM SOCKET IN MAIN: " + this.state.testObject);
       */ 
        

         //WEBSOCKET: WebSocket related
            /*const ws = new WebSocket(url);
            this.setState({webSocket: ws});

            ws.onopen = () => {
                console.log('Connected from client!');
            };

            ws.onmessage = message => {
                const text = message.data;
                this.setState({
                    messages: [...this.state.messages, text],
                });
            };*/ 


            /**
             * 
             * 
             * <Button 
                            color="danger" 
                            onClick={this.socketEvent}>SocketEvent
                        </Button>  
             */