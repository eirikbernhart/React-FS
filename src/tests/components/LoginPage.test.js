import React from 'react';
import {shallow} from 'enzyme';
import HeaderComponent from '../../components/HeaderComponent';
import toJSON from 'enzyme-to-json';
import { LoginPage } from '../../components/LoginPage';



test('should', () => {
    const wrapper = shallow(<LoginPage />);
    expect(wrapper).toMatchSnapshot();
});