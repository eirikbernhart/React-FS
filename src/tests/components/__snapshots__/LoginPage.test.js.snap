// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoginPage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="App-header"
        >
          <h2>
            Login
          </h2>
        </div>,
        <Form
          tag="form"
        >
          <FormGroup
            tag="div"
          >
            <Label
              for="userName"
              tag="label"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <strong>
                Username
              </strong>
            </Label>
            <Input
              id="userName"
              name="username"
              onChange={[Function]}
              placeholder="Username..."
              tag="p"
              type="username"
              value=""
            />
          </FormGroup>
          <FormGroup
            tag="div"
          >
            <Label
              for="examplePassword"
              tag="label"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <strong>
                Password
              </strong>
            </Label>
            <Input
              id="examplePassword"
              name="password"
              onChange={[Function]}
              placeholder="Password..."
              tag="p"
              type="password"
              value=""
            />
          </FormGroup>
          <Button
            color="danger"
            onClick={[Function]}
            tag="button"
          >
            Log in
          </Button>
        </Form>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h2>
            Login
          </h2>,
          "className": "App-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login",
          },
          "ref": null,
          "rendered": "Login",
          "type": "h2",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <FormGroup
              tag="div"
            >
              <Label
                for="userName"
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <strong>
                  Username
                </strong>
              </Label>
              <Input
                id="userName"
                name="username"
                onChange={[Function]}
                placeholder="Username..."
                tag="p"
                type="username"
                value=""
              />
            </FormGroup>,
            <FormGroup
              tag="div"
            >
              <Label
                for="examplePassword"
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <strong>
                  Password
                </strong>
              </Label>
              <Input
                id="examplePassword"
                name="password"
                onChange={[Function]}
                placeholder="Password..."
                tag="p"
                type="password"
                value=""
              />
            </FormGroup>,
            <Button
              color="danger"
              onClick={[Function]}
              tag="button"
            >
              Log in
            </Button>,
          ],
          "tag": "form",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Label
                  for="userName"
                  tag="label"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <strong>
                    Username
                  </strong>
                </Label>,
                <Input
                  id="userName"
                  name="username"
                  onChange={[Function]}
                  placeholder="Username..."
                  tag="p"
                  type="username"
                  value=""
                />,
              ],
              "tag": "div",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <strong>
                    Username
                  </strong>,
                  "for": "userName",
                  "tag": "label",
                  "widths": Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Username",
                  },
                  "ref": null,
                  "rendered": "Username",
                  "type": "strong",
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "id": "userName",
                  "name": "username",
                  "onChange": [Function],
                  "placeholder": "Username...",
                  "tag": "p",
                  "type": "username",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Label
                  for="examplePassword"
                  tag="label"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <strong>
                    Password
                  </strong>
                </Label>,
                <Input
                  id="examplePassword"
                  name="password"
                  onChange={[Function]}
                  placeholder="Password..."
                  tag="p"
                  type="password"
                  value=""
                />,
              ],
              "tag": "div",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <strong>
                    Password
                  </strong>,
                  "for": "examplePassword",
                  "tag": "label",
                  "widths": Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Password",
                  },
                  "ref": null,
                  "rendered": "Password",
                  "type": "strong",
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "id": "examplePassword",
                  "name": "password",
                  "onChange": [Function],
                  "placeholder": "Password...",
                  "tag": "p",
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Log in",
              "color": "danger",
              "onClick": [Function],
              "tag": "button",
            },
            "ref": null,
            "rendered": "Log in",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="App-header"
          >
            <h2>
              Login
            </h2>
          </div>,
          <Form
            tag="form"
          >
            <FormGroup
              tag="div"
            >
              <Label
                for="userName"
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <strong>
                  Username
                </strong>
              </Label>
              <Input
                id="userName"
                name="username"
                onChange={[Function]}
                placeholder="Username..."
                tag="p"
                type="username"
                value=""
              />
            </FormGroup>
            <FormGroup
              tag="div"
            >
              <Label
                for="examplePassword"
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <strong>
                  Password
                </strong>
              </Label>
              <Input
                id="examplePassword"
                name="password"
                onChange={[Function]}
                placeholder="Password..."
                tag="p"
                type="password"
                value=""
              />
            </FormGroup>
            <Button
              color="danger"
              onClick={[Function]}
              tag="button"
            >
              Log in
            </Button>
          </Form>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h2>
              Login
            </h2>,
            "className": "App-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Login",
            },
            "ref": null,
            "rendered": "Login",
            "type": "h2",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <FormGroup
                tag="div"
              >
                <Label
                  for="userName"
                  tag="label"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <strong>
                    Username
                  </strong>
                </Label>
                <Input
                  id="userName"
                  name="username"
                  onChange={[Function]}
                  placeholder="Username..."
                  tag="p"
                  type="username"
                  value=""
                />
              </FormGroup>,
              <FormGroup
                tag="div"
              >
                <Label
                  for="examplePassword"
                  tag="label"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <strong>
                    Password
                  </strong>
                </Label>
                <Input
                  id="examplePassword"
                  name="password"
                  onChange={[Function]}
                  placeholder="Password..."
                  tag="p"
                  type="password"
                  value=""
                />
              </FormGroup>,
              <Button
                color="danger"
                onClick={[Function]}
                tag="button"
              >
                Log in
              </Button>,
            ],
            "tag": "form",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <Label
                    for="userName"
                    tag="label"
                    widths={
                      Array [
                        "xs",
                        "sm",
                        "md",
                        "lg",
                        "xl",
                      ]
                    }
                  >
                    <strong>
                      Username
                    </strong>
                  </Label>,
                  <Input
                    id="userName"
                    name="username"
                    onChange={[Function]}
                    placeholder="Username..."
                    tag="p"
                    type="username"
                    value=""
                  />,
                ],
                "tag": "div",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": <strong>
                      Username
                    </strong>,
                    "for": "userName",
                    "tag": "label",
                    "widths": Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ],
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Username",
                    },
                    "ref": null,
                    "rendered": "Username",
                    "type": "strong",
                  },
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "id": "userName",
                    "name": "username",
                    "onChange": [Function],
                    "placeholder": "Username...",
                    "tag": "p",
                    "type": "username",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <Label
                    for="examplePassword"
                    tag="label"
                    widths={
                      Array [
                        "xs",
                        "sm",
                        "md",
                        "lg",
                        "xl",
                      ]
                    }
                  >
                    <strong>
                      Password
                    </strong>
                  </Label>,
                  <Input
                    id="examplePassword"
                    name="password"
                    onChange={[Function]}
                    placeholder="Password..."
                    tag="p"
                    type="password"
                    value=""
                  />,
                ],
                "tag": "div",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": <strong>
                      Password
                    </strong>,
                    "for": "examplePassword",
                    "tag": "label",
                    "widths": Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ],
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Password",
                    },
                    "ref": null,
                    "rendered": "Password",
                    "type": "strong",
                  },
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "id": "examplePassword",
                    "name": "password",
                    "onChange": [Function],
                    "placeholder": "Password...",
                    "tag": "p",
                    "type": "password",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Log in",
                "color": "danger",
                "onClick": [Function],
                "tag": "button",
              },
              "ref": null,
              "rendered": "Log in",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
